Add stats for pilot

dancunningham opened the pull request:
Adds some additional tools to the "dashboard" for the pilot team to use:
- Soft shutdown link
- Uptime stats and downtime report
- Backup and restore status page (including USB stick status)

To deploy this on a live server, all the files need to be put in place in Debian and then the commands in /packages/buendia-utils/control/postinst need to be run.

Alternatively, you could build new versions of the buendia-dashboard and buendia-utils packages and let the autoupdate script install those from USB.

llvasconcellos commented on packages/buendia-dashboard/data/usr/share/buendia/dashboard/index:
```diff
@@ -42,7 +53,7 @@ Content-Type: text/html
 <h1>Project Buendia server</h1>
 
-<h2>Staff access</h2>
+<h2>Bokoro Pilot - Server 9 (ATFC Simple Server)</h2>
```
This could have a TODO to make it dynamic and not hard coded.

llvasconcellos commented:
My only concern is the shutdown request check interval. Everything else seems fine.

dancunningham commented:
Thanks @llvasconcellos. I've changed the interval for the shutdown and reboot services to 30s, and added a TODO as you suggested, also removing the specific server name which probably shouldn't be committed.

capnfabs commented:
The dashboard stuff looks good, but the shutdown checker and the reboot checker should be unified - currently we've got a service in Debian checking for shutdown and a service in Yocto checking for reboots.

capnfabs commented:
Also, the fact that so many different pieces of code have to be able to interpret status codes is pretty scary.

dancunningham commented:
@capnfabs - both checks are on the Yocto side, I did the shutdown check in the same way as the reboot check. Do you recommend merging them both into a single service that checks for both reboot and shutdown requests then?

capnfabs commented:
@dancunningham Ah yes, you're right - sorry about that.

dancunningham commented:
@capnfabs @llvasconcellos does anything need done to this before we can merge it in?

I'd like to verify that it works on an Edison. For that we need to merge it and do a release anyway!

llvasconcellos commented:
@dancunningham It's fine by me. You can merge it.

capnfabs commented on packages/buendia-dashboard/data/usr/share/buendia/dashboard/backups:
```diff
@@ -0,0 +1,92 @@
+
+buendia_status=$(cat /home/buendia/server_status.txt)
+
+if [ $buendia_status -eq 2 ]; then
```
You can use bash `case` statements for this: http://tldp.org/LDP/Bash-Beginners-Guide/html/sect_07_03.html.

Also, as is, you should be quoting the `$buendia_status` variable in these tests because you're using `[` as opposed to `[[`. See http://serverfault.com/a/52050/145418

capnfabs commented on packages/buendia-dashboard/data/usr/share/buendia/dashboard/backups:
```diff
@@ -0,0 +1,92 @@
+successful_backups=$(cat -v /var/log/buendia/buendia-backup.log | grep '(end, status 0)' | sed -e 's/^/<li>/' -e 's/:[0-9][0-9]\W*[0-9]*> .*/<\/li>/')
+
+num_restores_done=$(cat -v /var/log/buendia/buendia-restore.log | grep '(end, status 0)' | wc -l)
+successful_restores=$(cat -v /var/log/buendia/buendia-restore.log | grep '(end, status 0)' | sed -e 's/^/<li>/' -e 's/:[0-9][0-9]\W*[0-9]*> .*/<\/li>/')
```
Some comments indicating what these regular expressions are doing would be helpful.

capnfabs commented on packages/buendia-dashboard/data/usr/share/buendia/dashboard/downtime:
```diff
@@ -0,0 +1,42 @@
+# OR CONDITIONS OF ANY KIND, either express or implied.  See the License for
+# specific language governing permissions and limitations under the License.
+
+up_count=$(cat /var/log/buendia/buendia-warmup.log | grep '(end, status 0)' | wc -l)
```
I'm not sure that `buendia-warmup.log` is a reliable source of data. `buendia-warmup` seems to run multiple times in parallel some times and I thought it truncated the logs from time to time. Having said, I prefer this to performing yet another health check :)

capnfabs commented on packages/buendia-dashboard/data/usr/share/buendia/dashboard/downtime:
```diff
@@ -0,0 +1,42 @@
+down_count=$(cat /var/log/buendia/buendia-warmup.log | grep '(end, status 1)' | wc -l)
+downtimes=$(cat /var/log/buendia/buendia-warmup.log | grep '(end, status 1)' | sed -e 's/:[0-9][0-9]\W*[0-9]*> .*/<br\/>/')
+
+# I would just use bc but it's not installed on Debian so we're constrained to integer arithmetic
```
You can install `bc` if you want. If you don't want to, a better comment would be "It's not worth installing `bc` just for this calculation so we constrain this calculation to integer arithmetic".

capnfabs commented on packages/buendia-dashboard/data/usr/share/buendia/dashboard/index:
```diff
@@ -17,6 +17,17 @@ tail -5 /var/log/buendia/buendia-warmup.log | grep -q Success && updown=SERVING
 
 uptime=$(uptime | sed -e 's/.* up//' -e 's/, *[0-9][0-9]* *user.*//')
 
+up_count=$(cat /var/log/buendia/buendia-warmup.log | grep '(end, status 0)' | wc -l)
```
Please refactor these into a common script / location so that if we need to modify the logic we only need to modify it in one place.

capnfabs commented:
All in all, looks good, but back over to you for adjustments and tweaks @dancunningham :)

zestyping commented:
This PR has been reconstituted in #117.
